src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -253,27 +254,97 @@ ExpressionEvaluator::_evaluate_in_expression<ExpressionEvaluator::Bool>(const In
+        }
+
+        if (element->type != ExpressionType::Value) {
+          all_elements_are_values_of_left_type = false;
+        } else {
+          const auto& value_expression = std::static_pointer_cast<ValueExpression>(element);
+          if (value_expression->value.type() != typeid(LeftDataType)) all_elements_are_values_of_left_type = false;
+        }
       }
-    }
+    });
 
     if (type_compatible_elements.empty()) {
-      // `5 IN ()` is FALSE as is `NULL IN ()`
+      // `5 IN ()` is FALSE
       return std::make_shared<ExpressionResult<ExpressionEvaluator::Bool>>(std::vector<ExpressionEvaluator::Bool>{0});
     }
 
+    // If all elements of the list are simple values (e.g., `IN (1, 2, 3)`), iterate over the column and compare the
+    // values. Otherwise, translate the IN clause to a series of ORs:
+    // "a IN (x, y, z)"   ---->   "a = x OR a = y OR a = z"
+    if (all_elements_are_values_of_left_type) {
+      _resolve_to_expression_result_view(left_expression, [&](const auto& view) {
+        using LeftDataType = typename std::decay_t<decltype(view)>::Type;
+
+        // Above, we have ruled out NULL on the left side, but the compiler does not know this yet
+        if constexpr (!std::is_same_v<LeftDataType, NullValue>) {
+          std::vector<LeftDataType> right_values;
+          right_values.reserve(type_compatible_elements.size());
+          for (const auto& expression : type_compatible_elements) {
+            const auto& value_expression = std::static_pointer_cast<ValueExpression>(expression);
+            right_values.emplace_back(boost::get<LeftDataType>(value_expression->value));
+          }
+          std::sort(right_values.begin(), right_values.end());
+
+          result_values.resize(view.size());
+          if (left_expression.is_nullable()) {
+            result_nulls.resize(view.size());
+            for (auto chunk_offset = ChunkOffset{0}; chunk_offset < view.size(); ++chunk_offset) {
+              if (view.is_null(chunk_offset)) {
+                result_nulls[chunk_offset] = true;
+                continue;
+              }
+              if (auto it = std::lower_bound(right_values.cbegin(), right_values.cend(), view.value(chunk_offset));
+                  it != right_values.cend() && *it == view.value(chunk_offset)) {
+                result_values[chunk_offset] = true;
+              }
+            }
+          } else {
+            for (auto chunk_offset = ChunkOffset{0}; chunk_offset < view.size(); ++chunk_offset) {
+              if (auto it = std::lower_bound(right_values.cbegin(), right_values.cend(), view.value(chunk_offset));
+                  it != right_values.cend() && *it == view.value(chunk_offset)) {
+                result_values[chunk_offset] = true;
+              }
+            }
+          }
+        } else {
+          Fail("Should have ruled out NullValues on the left side of IN by now");
+        }
+      });
+
+      if (left_expression.is_nullable()) {
```

Comment:
No need for this if block, just pass in an empty nulls vector in the non-nullable case, the ExpressionResult is designed to handle this.

Reply:
neat